### ISSUE TITLE
Corrupt data file after {:error, :enomem}

### ISSUE THREAD

**Issue by vorce** [lucaong/cubdb#15]
Just ran in to this error. Basically it looks like my app ran out of memory (for some yet unknown reason), this caused it to crash. After trying to start the application again CubDB cannot start.

Initial stacktrace:
```
16:40:11.775 [error] Task :cubdb started from #PID<0.3750.0> terminating
 ** (WithClauseError) no with clause matching: {:error, :enomem}
     (cubdb 0.17.0) lib/cubdb/store/file.ex:94: CubDB.Store.CubDB.Store.File.get_node/2
     (cubdb 0.17.0) lib/cubdb/btree.ex:337: CubDB.Btree.lookup_leaf/4
     (cubdb 0.17.0) lib/cubdb/btree.ex:106: CubDB.Btree.fetch/2
     (cubdb 0.17.0) lib/cubdb/reader.ex:31: CubDB.Reader.run/4
     (elixir 1.10.2) lib/task/supervised.ex:90: Task.Supervised.invoke_mfa/2
     (stdlib 3.11.2) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
```

Error when starting again
```
** (Mix) Could not start application myapp: MyApp.Application.start(:normal, []) returned an error: shutdown: failed to start child: CubDB
    ** (EXIT) an exception was raised:
        ** (ArgumentError) argument error
            (cubdb) lib/cubdb/store/file.ex:94: CubDB.Store.CubDB.Store.File.get_node/2
            (cubdb) lib/cubdb/btree.ex:64: CubDB.Btree.new/2
            (cubdb) lib/cubdb.ex:786: CubDB.init/1
            (stdlib) gen_server.erl:374: :gen_server.init_it/2
            (stdlib) gen_server.erl:342: :gen_server.init_it/6
            (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
```

Seems like the data file has been corrupt or something? Any ideas on how to get to a working state without data loss?

**Comment by lucaong** [comment]
Hi @vorce , thanks for reporting this.

The strange thing is that the app ran out of memory when reading, not when writing, so the data file should not have been corrupted. It is possible that something else corrupted the file, which then caused the out of memory error.

Can I ask you more info about your system? Is it an embedded device or a computer? What kind of operations are you performing?

**Comment by lucaong** [comment]
If you still have the file, it could be interesting to inspect it.

I did not witness this kind of file corruption, so it is valuable for me to investigate it. The only other known case of file corruption that I encountered was on an embedded device, as the result of faulty power supply (voltage too low, causing random bit flipping when writing on the SD card, unfortunately not something that CubDB can protect against).

If the file corruption happened upon the last write before the out-of-memory error, you could try to create a copy of the database file truncating the last part of the file, until you find a point in which `CubDB` is able to open the file. The problematic write would be lost, but the rest would be still available. The strange thing is that `CubDB` should be able to recover automatically from these kinds of problems, therefore if the file does not contain private data you could try to send it to me for inspection.

**Comment by vorce** [comment]
Hey @lucaong!
This happened on a Digital Ocean droplet (cheapest ["Standard" config](https://www.digitalocean.com/pricing/)). What I did at the time was concurrent reads if I recall correctly.

I still have the file, attaching it. Thanks for the response, and a great project. Let me know if you need more information.

[4.cub.crashing.zip](https://github.com/lucaong/cubdb/files/4440432/4.cub.crashing.zip)

**Comment by lucaong** [comment]
Hi @vorce,
the most recent header in the file points to the wrong location. That could be caused by multiple `CubDB` processes being open on the same data directory and writing to it (for example if you run your app in multiple instances, or if you run the app in a `iex` console on the same machine where your app is running). Only one `CubDB` process should use a given data dir at any time. Do you think that could be the case?

I will investigate more, but in the meanwhile, here is your file fixed by truncating the latest header:

[4.cub.fixed.zip](https://github.com/lucaong/cubdb/files/4443684/4.cub.fixed.zip)

**Comment by vorce** [comment]
@lucaong 
That sounds exactly what happened! I ran two docker contains with my app, both mounting a dir on the host to access the cubdb file.

Since then I am only running one instance. But If I need to scale up my app I can't use cubdb anymore?

I think we can close this issue. But would be happy to continue the discussion if you have any other input. Thanks again!

**Comment by lucaong** [comment]
Hi @vorce,
yes, `CubDB` is an embedded database: it means that it runs "inside" your app instance, as opposed to as a separate service (like Postgres). This is useful in some applications (like embedded devices, or desktop apps), but does not work for web apps that can run in multiple instances, unless you have a case in which each instance has its own separate data store.

Sorry for the confusion :) I tried to make it clearer in the [FAQ section](https://hexdocs.pm/cubdb/faq.html#why-should-i-use-cubdb), but I can see how it can be confusing when coming from different databases.

I'd like to find a good way to alert users when mistakenly opening the same data directory from multiple `CubDB` processes, but I did not find yet a way that is satisfactory to me (`CubDB` will raise an error when starting multiple processes on the same data directory _from the same node_, but not yet across nodes).
